### ISSUE TITLE
Function call result materialization is failing

### ISSUE THREAD

**Issue by vladascibulskis** [OData/odata.net#2236]
I'm running an OData service that has unbound OData function.
Function returns following result:

```
{
"FirstName":"name",
"LastName":"surname",
"UserName":"username",
"GroupNames":[
    "groupName1",
    "gropuName2"
  ]
}
```

I'm calling this function using OData.Client as follows:

`this.odataApi.WhoAmI().GetValue();`

And get following error:

[13:56:54 ERR] First chance exception occured
System.NullReferenceException: Object reference not set to an instance of an object.
   at Microsoft.OData.Client.Materialization.CollectionValueMaterializationPolicy.ApplyCollectionDataValues(ODataProperty collectionProperty, Object collectionInstance, Type collectionItemType, Action`2 addValueToBackingICollectionInstance, Boolean isElementNullable)

### Assemblies affected

    <PackageReference Include="Microsoft.OData.Client" Version="7.9.3" />

### Reproduce steps

Create OData unbound function witch returns object with IEnumerable<string> as one of it's parameters (in my case it's list of user groups)
Generate client proxy.
Call this function.

### Expected result

Expecting to get correctly materialized result.

### Actual result

Exception.

**Comment by gathogojr** [comment]
@vladascibulskis Thank you for reporting the issue. Unfortunately I was unable to reproduce it. Kindly take a look at my attempt at a repro [here](https://github.com/gathogojr/ODataNetIssue2236Repro). Materialization worked okay

**Comment by vladascibulskis** [comment]
@gathogojr Thank you for looking at it. I tried to reproduce the issue in your repository and could not do that.
I have generated proxy classes for my own service and it also worked fine.

I have regenerated proxy for my project and it also works now, so this issue can be closed.

After additional testing I now believe the issue was related to outdated ...Csdl.xml file. For some reason it wasn't updated and still had GroupName as simple string type instead of Collection. 
I'm just surprised that xml file is used in materialization...

**Comment by gathogojr** [comment]
@vladascibulskis Glad to have been of some help. The xml file is not used during materialization. Regenerating the proxy might have fixed the issue

**Comment by vladascibulskis** [comment]
At some point it actually depends on xml.
I was able to reproduce the issue in your repo with following change:

![image](https://user-images.githubusercontent.com/16113774/141176186-ff27a355-772b-4ba8-bbde-696a3090fb1e.png)

I'm guessing the main issue is the name change from plural to singular form, and it probably fails to deserialize the JSON response. Might consider if you want to add better error in that case, but not sure if it's common enough to have specific handling...

**Comment by vladascibulskis** [comment]
@gathogojr One more additional note on how I got into this situation:
I set "IncludeT4File": true  in  ConnectedService.json and modified the templates for my own purposes. So when I need to refresh my proxy, I simply run the T4 template and don't do the update on the Connected service (because it would overwrite my T4 modifications).

As I have multiple projects using same OData service, I simply copied T4 template files from other project and the template had incorrect path for .xml file.
This resulted in the fact that xml file was being updated in different location and not in this project.


Based on the property description in T4 template:

> 	// The path for the temporary file where the metadata xml document can be stored. Use this if your metadata is too big to be stored in a string literal. Ensure that you have write permission for this path.
> 	// For example - "C:\\temp\\Test.xml"
> 	public const string MetadataFilePath = @"D:\path\ListenerClient\ListenerClientCsdl.xml";
> 

I would expect that this file should not be used during runtime, but based on the fact that changing it does cause the exception, I believe it's used somewhere and that's probably a bug worth fixing.
